crates/wasmtime/src/runtime/component/instance.rs

```diff
@@ -547,23 +547,42 @@ impl<'a> Instantiator<'a> {
         store.modules_mut().register_component(component);
         let imported_resources: ImportedResources =
             PrimaryMap::with_capacity(env_component.imported_resources.len());
+        let instance = Instance(store.store_data_mut().insert(None));
+        store[instance.0] = Some(Box::new(InstanceData {
+            instances: PrimaryMap::with_capacity(env_component.num_runtime_instances as usize),
+            component: component.clone(),
+            state: OwnedComponentInstance::new(
+                component.runtime_info(),
+                Arc::new(imported_resources),
+                store.traitobj(),
+                instance,
+            ),
+            imports: imports.clone(),
+        }));
+
         Instantiator {
             component,
             imports,
             core_imports: OwnedImports::empty(),
-            data: InstanceData {
-                instances: PrimaryMap::with_capacity(env_component.num_runtime_instances as usize),
-                component: component.clone(),
-                state: OwnedComponentInstance::new(
-                    component.runtime_info(),
-                    Arc::new(imported_resources),
-                    store.traitobj(),
-                ),
-                imports: imports.clone(),
-            },
+            instance,
         }
     }
 
+    fn data<'b>(&self, store: &'b mut StoreOpaque) -> &'b mut InstanceData {
+        store[self.instance.0].as_mut().unwrap()
+    }
+
+    fn with_data<R>(
+        &self,
+        store: &mut StoreOpaque,
+        fun: impl FnOnce(&mut StoreOpaque, &mut InstanceData) -> R,
+    ) -> R {
+        let mut data = store[self.instance.0].take().unwrap();
+        let result = fun(store, &mut data);
+        store[self.instance.0] = Some(data);
```

Comment:
Is this the bit where if `fun` panics it leaks `data`? If so could that be solved with an RAII guard?